resteasy-client/src/main/java/org/jboss/resteasy/client/jaxrs/engines/JettyClientEngine.java

```diff
@@ -0,0 +1,223 @@
+        } catch (ExecutionException | TimeoutException e) {
+            throw clientException(e.getCause(), null); // NOPMD
+        }
+    }
+
+    /**
+     * Implementation note: due to lack of asynchronous message decoders the request must either be buffered,
+     * or it must have a {@code null} extractor and type parameter {@code <T>} must be {@link ClientResponse},
+     * which will read the data through its stream.  It is not possible to use the synchronous JAX-RS message
+     * decoding infrastructure without buffering or spinning up auxiliary threads (arguably more expensive than buffering).
+     *
+     * @see AsyncClientHttpEngine#submit(ClientInvocation, boolean, InvocationCallback, org.jboss.resteasy.client.jaxrs.AsyncClientHttpEngine.ResultExtractor)
+     */
+    @SuppressWarnings("resource")
+    @Override
+    public <T> Future<T> submit(ClientInvocation invocation, boolean buffered, InvocationCallback<T> callback, ResultExtractor<T> extractor) {
+        final Request request = client.newRequest(invocation.getUri());
+        final CompletableFuture<T> future = new RequestFuture<T>(request);
+
+        request.method(invocation.getMethod());
+        invocation.getHeaders().asMap().forEach((h, vs) -> vs.forEach(v -> request.header(h, v)));
+
+        final DeferredContentProvider content;
+        if (invocation.getEntity() != null) {
+            content = new DeferredContentProvider();
+            request.content(content);
+        } else {
+            content = null;
+        }
+
+        request.send(new Response.Listener.Adapter() {
+            private ClientResponse cr;
+            private JettyResponseStream stream = new JettyResponseStream();
+
+            @Override
+            public void onHeaders(Response response) {
+                cr = new JettyClientResponse(invocation.getClientConfiguration(), stream, () -> future.cancel(true));
+                cr.setProperties(invocation.getMutableProperties());
+                cr.setStatus(response.getStatus());
+                cr.setHeaders(extract(response.getHeaders()));
+                if (!buffered) {
+                    if (extractor != null) {
+                        throw new IllegalStateException("TODO: ResultExtractor is synchronous and may not be used without buffering.");
+                    }
+                    future.complete((T) cr); // TODO: dangerous cast, see javadoc!
+                }
+            }
+
+            @Override
+            public void onContent(Response response, ByteBuffer content, Callback callback) {
+                stream.offer(content, callback);
+            }
+
+            @Override
+            public void onFailure(Response response, Throwable failure) {
+                future.completeExceptionally(failure);
+                callback.failed(failure);
+            }
+
+            @Override
+            public void onSuccess(Response response) {
+                if (buffered) {
```

Comment:
In the case that we're unbuffered, we'll have a `null` extractor, so is that why it's okay to do nothing here in that case?

Reply:
Yup, exactly.